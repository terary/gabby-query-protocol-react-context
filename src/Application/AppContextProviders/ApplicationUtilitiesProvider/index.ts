import ApplicationUtilitiesProvider, {
  ApplicationUtilitiesContext,
  useTestUtilities,
} from "./ApplicationUtilitiesProvider";
import type { ApplicationUtilitiesType } from "./type";

export default ApplicationUtilitiesProvider;
export { ApplicationUtilitiesContext, useTestUtilities };
export type { ApplicationUtilitiesType };
