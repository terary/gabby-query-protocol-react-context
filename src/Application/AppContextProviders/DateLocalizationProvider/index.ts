import DateLocalizationProvider from "./Provider";

export default DateLocalizationProvider;
