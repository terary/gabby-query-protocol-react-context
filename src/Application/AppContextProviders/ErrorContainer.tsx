import { ReactNode } from "react";
import { ErrorBoundary } from "react-error-boundary";

function ErrorFallback({ error }: { error: Error }) {
  return (
    <div role="alert">
      <p>Something went wrong:</p>
      <pre>{error.message}</pre>
    </div>
  );
}

export const ErrorContainer = ({ children }: { children: ReactNode }) => {
  return <ErrorBoundary FallbackComponent={ErrorFallback}>{children}</ErrorBoundary>;
};
