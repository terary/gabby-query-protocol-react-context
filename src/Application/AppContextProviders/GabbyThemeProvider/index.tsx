import { GabbyThemeProvider } from "./Provider";
export default GabbyThemeProvider;
