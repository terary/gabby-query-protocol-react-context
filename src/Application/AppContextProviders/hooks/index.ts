import { useApplicationUtilities } from "./useApplicationUtilities";

export { useApplicationUtilities };
