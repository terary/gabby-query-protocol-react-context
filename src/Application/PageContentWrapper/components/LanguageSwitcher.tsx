import * as React from "react";

import { SUPPORTED_LOCALE_DICTIONARY, useApplicationUtilities } from "../..";
import type { TSupportedLocales } from "../..";
import { useTheme, makeStyles } from "@mui/material/styles";

import MenuItem from "@mui/material/MenuItem";
import FormControl from "@mui/material/FormControl";
import Select, { SelectChangeEvent } from "@mui/material/Select";

export const LanguageSwitcher = () => {
  const { setCurrentLocale, getCurrentLocaleCode, getLocale, t } = useApplicationUtilities();
  const theme = useTheme();
  const handleChange = (event: SelectChangeEvent) => {
    const localeCode = event.target.value as TSupportedLocales;
    const localeSettings = getLocale(localeCode);
    setCurrentLocale(event.target.value as TSupportedLocales);

    const htmlPage = document.querySelector("html");
    if (htmlPage !== null) {
      htmlPage.setAttribute("dir", localeSettings.languageDir);
      htmlPage.setAttribute("lang", localeSettings.languageCode);
    }
  };

  return (
    <div>
      <FormControl variant="standard" sx={{ m: 1, minWidth: 120 }}>
        {/* <InputLabel id="demo-simple-select-standard-label">Age</InputLabel> */}
        <Select
          labelId="demo-simple-select-standard-label"
          id="demo-simple-select-standard"
          value={getCurrentLocaleCode()}
          onChange={handleChange}
          label={t("Language")}
          sx={{ color: theme.palette.grey[100], borderBottom: "1px solid white" }}
        >
          {Object.entries(SUPPORTED_LOCALE_DICTIONARY).map(([localeKey, localeSetting]) => {
            {
              return (
                <MenuItem value={localeKey} key={localeKey}>
                  {localeSetting.countryNameLocal}
                </MenuItem>
              );
            }
          })}
        </Select>
      </FormControl>
      {/* <FormControl variant="filled" sx={{ m: 1, minWidth: 120 }}>
        <InputLabel id="demo-simple-select-filled-label">Age</InputLabel>
        <Select
          labelId="demo-simple-select-filled-label"
          id="demo-simple-select-filled"
          value={age}
          onChange={handleChange}
        >
          <MenuItem value="">
            <em>None</em>
          </MenuItem>
          <MenuItem value={10}>Ten</MenuItem>
          <MenuItem value={20}>Twenty</MenuItem>
          <MenuItem value={30}>Thirty</MenuItem>
        </Select>
      </FormControl> */}
    </div>
  );
};
