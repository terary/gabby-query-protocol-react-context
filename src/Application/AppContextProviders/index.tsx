import React from "react";
import { TSupportedLocales } from "..";
import ApplicationUtilitiesProvider, {
  useTestUtilities,
} from "./ApplicationUtilitiesProvider";
import { ErrorContainer } from "./ErrorContainer";
import DateLocalizationProvider from "./DateLocalizationProvider";
import GabbyThemeProvider from "./GabbyThemeProvider";

interface IAppProviderParameters {
  children?: React.ReactNode;
  initialLocaleCode?: TSupportedLocales;
}

export const AppContextProviders = ({
  initialLocaleCode,
  children,
}: IAppProviderParameters) => {
  return (
    <ErrorContainer>
      {" "}
      {/* Error container may have issue with hooks, requires testing */}
      <ApplicationUtilitiesProvider initialLocaleCode={initialLocaleCode}>
        <GabbyThemeProvider>
          <DateLocalizationProvider>{children}</DateLocalizationProvider>
        </GabbyThemeProvider>
      </ApplicationUtilitiesProvider>
    </ErrorContainer>
  );
};

export { useTestUtilities };
