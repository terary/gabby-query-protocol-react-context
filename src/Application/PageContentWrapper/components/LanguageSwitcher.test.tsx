import { LanguageSwitcher } from "./LanguageSwitcher";
import renderer, { act } from "react-test-renderer";
import { render, fireEvent, within, cleanup } from "@testing-library/react";
import { toMatchDiffSnapshot } from "snapshot-diff";
import { AppContextProviders } from "../../AppContextProviders";
import { useApplicationUtilities } from "../../index";
expect.extend({ toMatchDiffSnapshot });

const LanguageSwitcherWrapper = () => {
  const { getCurrentLocaleCode } = useApplicationUtilities();
  return (
    <div>
      <LanguageSwitcher />
      current language:
      <span data-testid="the-current-language-code">{getCurrentLocaleCode()}</span>
    </div>
  );
};

const LanguageSwitcherAppWrapper = () => {
  return (
    <AppContextProviders>
      <LanguageSwitcherWrapper />
    </AppContextProviders>
  );
};

describe("LanguageSwitcher", () => {
  afterEach(cleanup);
  it("Should render", () => {
    const tree = renderer.create(<LanguageSwitcherAppWrapper />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should change language", () => {
    const { getByTestId, getByRole } = render(<LanguageSwitcherAppWrapper />);

    expect(getByTestId("the-current-language-code")).toBeInTheDocument();
    expect(getByTestId("the-current-language-code").innerHTML).toBe("es_mx");

    fireEvent.mouseDown(getByRole("button"));
    const listbox = within(getByRole("listbox"));
    fireEvent.click(listbox.getByText(/Deutsch/i));

    fireEvent.click(getByRole("button"));

    expect(getByTestId("the-current-language-code").innerHTML).toBe("de_de");
    cleanup(); // this doesn't re-set app state
  });
  it("Should change language direction accordingly", () => {
    const { getByTestId, getByRole, container } = render(<LanguageSwitcherAppWrapper />);

    expect(getByTestId("the-current-language-code")).toBeInTheDocument();
    // expect(getByTestId("the-current-language-code").innerHTML).toBe("ex_mx");

    fireEvent.mouseDown(getByRole("button"));
    const listbox = within(getByRole("listbox"));
    fireEvent.click(listbox.getByText(/المغرب/i));
    fireEvent.click(getByRole("button"));

    expect(getByTestId("the-current-language-code").innerHTML).toBe("ar_ma");
    const htmlDoc = container.querySelector("html");
    expect(htmlDoc).toBeNull();
  });
});
