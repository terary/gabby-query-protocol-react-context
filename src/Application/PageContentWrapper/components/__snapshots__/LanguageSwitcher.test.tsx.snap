// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSwitcher Should render 1`] = `
Array [
  " ",
  <div
    className="Toastify"
  />,
  <div>
    <div>
      <div
        className="MuiFormControl-root css-1869usk-MuiFormControl-root"
      >
        <div
          className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiSelect-root css-163q0ur-MuiInputBase-root-MuiInput-root"
          onClick={[Function]}
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="demo-simple-select-standard-label demo-simple-select-standard"
            className="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
            id="demo-simple-select-standard"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="button"
            tabIndex={0}
          >
            MÃ©xico
          </div>
          <input
            aria-hidden={true}
            className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            disabled={false}
            onAnimationStart={[Function]}
            onChange={[Function]}
            required={false}
            tabIndex={-1}
            value="es_mx"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    current language:
    <span
      data-testid="the-current-language-code"
    >
      es_mx
    </span>
  </div>,
]
`;
